firefox-ios/firefox-ios-tests/Tests/ClientTests/BrowserViewControllerTests.swift

```diff
@@ -415,10 +419,84 @@ class BrowserViewControllerTests: XCTestCase, StoreTestUtility {
         XCTAssertEqual(browserCoordinator.showSummarizePanelCalled, 1)
     }
 
+    // MARK: - Record visit observation for History Panel
+
+    func testRecordVisitObservationIsCalledForNavigateInNewTabWithTitle() {
+        let subject = createSubject()
+        let tab = MockTab(profile: profile, windowUUID: .XCTestDefaultUUID)
+        let mockTabWebView = MockTabWebView(tab: tab)
+        let url = URL(string: "https://example.com/")
+        mockTabWebView.loadedURL = url
+        tab.webView = mockTabWebView
+
+        subject.navigateInTab(tab: tab, to: nil, webViewStatus: .title)
+
+        XCTAssertEqual(recordVisitManager.recordVisitCalled, 1)
+        XCTAssertNotNil(recordVisitManager.lastVisitObservation)
+    }
+
+    func testRecordVisitObservationIsCalledForNavigateInNewTabWithURL() {
+        let subject = createSubject()
+        let tab = MockTab(profile: profile, windowUUID: .XCTestDefaultUUID)
+        let mockTabWebView = MockTabWebView(tab: tab)
+        let url = URL(string: "https://example.com/")
+        mockTabWebView.loadedURL = url
+        tab.webView = mockTabWebView
+
+        subject.navigateInTab(tab: tab, to: nil, webViewStatus: .url)
+
+        XCTAssertEqual(recordVisitManager.recordVisitCalled, 1)
+        XCTAssertNotNil(recordVisitManager.lastVisitObservation)
+    }
+
+    func testRecordVisitObservationIsNotCalledForInternalURL() {
+        let subject = createSubject()
+        let tab = MockTab(profile: profile, windowUUID: .XCTestDefaultUUID)
+        let mockTabWebView = MockTabWebView(tab: tab)
+        let url = URL(string: "http://localhost:\(AppInfo.webserverPort)/")
+        mockTabWebView.loadedURL = url
+        tab.webView = mockTabWebView
+
+        subject.navigateInTab(tab: tab, to: nil, webViewStatus: .title)
+
+        XCTAssertEqual(recordVisitManager.recordVisitCalled, 0)
+        XCTAssertNil(recordVisitManager.lastVisitObservation)
+    }
+
+    func testRecordVisitObservationIsNotCalledForFileURL() {
+        let subject = createSubject()
+        let tab = MockTab(profile: profile, windowUUID: .XCTestDefaultUUID)
+        let mockTabWebView = MockTabWebView(tab: tab)
+        let url = URL(fileURLWithPath: "/tmp/test.html")
+        mockTabWebView.loadedURL = url
+        tab.webView = mockTabWebView
+
+        subject.navigateInTab(tab: tab, to: nil, webViewStatus: .title)
+
+        XCTAssertEqual(recordVisitManager.recordVisitCalled, 0)
+        XCTAssertNil(recordVisitManager.lastVisitObservation)
+    }
+
+    func testResetObservationIsCalledForAddNewTabAction() throws {
```

Comment:
This test catches a memory leak in `focusLocationTextField` from the addNewTab code path